replica/table.cc

```diff
@@ -562,10 +562,27 @@ void storage_group_manager::for_each_storage_group(std::function<void(size_t, st
     }
 }
 
+future<std::unique_ptr<storage_group>> storage_group_manager::remove_storage_group(size_t id) {
+    rwlock::holder exclusive_lock = co_await get_rwlock().hold_write_lock();
+
+    std::unique_ptr<storage_group> sg_ptr = std::move(_storage_groups[id]);
+    _storage_groups.erase(id);
+    if (!sg_ptr) {
+        throw std::out_of_range(format("remove_storage_group: storage group with id={} not found", id));
+    }
+
+    // storage_group_manager shouldn't see compaction groups, even though they aren't destroyed yet.
+    for (auto& cg_ptr : sg_ptr->compaction_groups()) {
```

Comment:
Hmm, I see this is done by the (sole) caller - cleanup_tablet.
It seems like it should be the one that acquires the exclusive lock
and this function could be synchronous and just do the low-level removal work.
It should be documented that the caller is responsible for stopping the removed storage_group's compaction groups.

Reply:
Otherwise, yielding after stopping the compaction groups and before the storage group is erased would wrongly give access the the stopped compaction groups, so stopping the compaction groups and erasing the the storage group must happen atomically for anyone accessing the storage_groups asynchronously, while holding the shared lock.